Add Taints and Tolerations

nicolethoen opened the pull request:
[CONSOLE-1275](https://jira.coreos.com/browse/CONSOLE-1275)
[Design](https://github.com/openshift/openshift-origin-design/pull/148)

**need to modify the dialog so existing taints/tolerations are read only.

![Mar-18-2019 10-44-42](https://user-images.githubusercontent.com/26305082/54538581-e17df480-496a-11e9-9710-ece1293779d4.gif)
![Mar-18-2019 10-44-50](https://user-images.githubusercontent.com/26305082/54538703-17bb7400-496b-11e9-9a21-a05e946c10ae.gif)

![Screen Shot 2019-03-18 at 10 20 11 AM](https://user-images.githubusercontent.com/26305082/54538610-ef337a00-496a-11e9-9a33-359c715d21dc.png)
![Screen Shot 2019-03-18 at 10 20 21 AM](https://user-images.githubusercontent.com/26305082/54538611-efcc1080-496a-11e9-89e5-6abe0e833e6d.png)
![Screen Shot 2019-03-18 at 10 29 19 AM](https://user-images.githubusercontent.com/26305082/54538612-efcc1080-496a-11e9-930c-df9a6a5900bc.png)
![Screen Shot 2019-03-18 at 10 46 11 AM](https://user-images.githubusercontent.com/26305082/54538689-112cfc80-496b-11e9-801f-5aab98a061ef.png)
![Screen Shot 2019-03-18 at 10 46 22 AM](https://user-images.githubusercontent.com/26305082/54538690-112cfc80-496b-11e9-8f3b-c0ddb7822f2d.png)

spadgett commented on frontend/public/components/modals/taint-update-modal.tsx:
```diff
@@ -0,0 +1,116 @@
+  _addRow = () => {
+    const clonedTaints = _.cloneDeep(this.state.taints);
+    clonedTaints.push({key: '', value: '', effect: ''});
+    this.setState({taints: clonedTaints});
```
This should use a `setState` callback. It's needed when updating state based on the previous value since `this.state` can be stale.

```ts
  this.setState((state: TaintUpdateModalState) => ({
    taints: [...state.taints, {key: '', value: '', effect: ''}],
  }));
```

https://reactjs.org/docs/react-component.html#setstate

spadgett commented on frontend/public/components/modals/taint-update-modal.tsx:
```diff
@@ -0,0 +1,116 @@
+  _change = (e, i, field) => {
+    const clonedTaints = _.cloneDeep(this.state.taints);
+    clonedTaints[i][field] = e.target ? e.target.value : e;
+    this.setState({taints: clonedTaints});
```
Use a `setState` callback

spadgett commented on frontend/public/components/modals/taint-update-modal.tsx:
```diff
@@ -0,0 +1,116 @@
+  };
+
+  _remove = (i) => {
+    const clonedTaints = _.cloneDeep(this.state.taints);
```
Use a `setState` callback

spadgett commented on frontend/public/components/modals/toleration-update-modal.tsx:
```diff
@@ -0,0 +1,129 @@
+    e.preventDefault();
+
+    // We just throw away any rows where the key is blank
+    const tolerations = _.reject(this.state.tolerations, t => _.isEmpty(t.key));
```
This has a special meaning. We can't just throw away empty keys.

> An empty key with operator Exists matches all keys, values and effects which means this will tolerate everything.

https://kubernetes.io/docs/concepts/configuration/taint-and-toleration/

spadgett commented on frontend/public/components/modals/toleration-update-modal.tsx:
```diff
@@ -0,0 +1,129 @@
+
+    const keys = tolerations.map(t => t.key);
+    if (_.uniq(keys).length !== keys.length) {
+      this.setState({errorMessage: 'Duplicate keys found.'});
```
Remove duplicate key check

spadgett commented on frontend/public/components/modals/toleration-update-modal.tsx:
```diff
@@ -0,0 +1,129 @@
+    }
+
+    // Convert any blank values to null
+    _.each(tolerations, t => t.value = _.isEmpty(t.value) ? null : t.value);
```
Tolerations that use the `Exists` operator will have an empty value. We can't just throw those away.

spadgett commented on frontend/public/components/modals/toleration-update-modal.tsx:
```diff
@@ -0,0 +1,129 @@
+  _change = (e, i, field) => {
+    const clonedTolerations = _.cloneDeep(this.state.tolerations);
+    clonedTolerations[i][field] = e.target ? e.target.value : e;
+    this.setState({tolerations: clonedTolerations});
```
Use a `setState` callback

spadgett commented on frontend/public/components/modals/toleration-update-modal.tsx:
```diff
@@ -0,0 +1,129 @@
+  _remove = (i) => {
+    const clonedTolerations = _.cloneDeep(this.state.tolerations);
+    clonedTolerations.splice(i, 1);
+    this.setState({tolerations: clonedTolerations});
```
Use a `setState` callback

spadgett commented on frontend/public/components/modals/toleration-update-modal.tsx:
```diff
@@ -0,0 +1,129 @@
+  _addRow = () => {
+    const clonedTolerations = _.cloneDeep(this.state.tolerations);
+    clonedTolerations.push({key: '', operator: 'Exists', value: '', effect: ''});
+    this.setState({tolerations: clonedTolerations});
```
Use a `setState` callback

spadgett commented on frontend/public/components/modals/toleration-update-modal.tsx:
```diff
@@ -0,0 +1,129 @@
+                selectedKey={c.operator} />
+            </div>
+            <div className="col-xs-3 pairs-list__name-field"><input type="text" className="form-control" value={c.value} onChange={(e) => this._change(e, i, 'value')} /></div>
+            <div className="col-xs-3 pairs-list__name-field"><input type="text" className="form-control" value={c.effect} onChange={(e) => this._change(e, i, 'effect')} /></div>
```
Should be a dropdown of values

spadgett commented on frontend/public/components/pod.jsx:
```diff
@@ -185,6 +187,8 @@ export const PodDetailsList = ({pod}) => {
     <dt>Node</dt>
     <dd><NodeLink name={pod.spec.nodeName} /></dd>
+    <dt>Tolerations</dt>
+    <dd><a className="co-m-modal-link" onClick={Kebab.factory.ModifyToleration(PodModel, pod).callback}>{pluralize(_.size(pod.spec.tolerations), 'Toleration')}</a></dd>
```
We should add this to other workload pages.

For pods, tolerations cannot be edited. So this should be a readonly modal.

spadgett commented on frontend/public/components/modals/_modals.scss:
```diff
@@ -90,3 +90,7 @@
   margin-top: 0;
 }
+
+.toleration-update-modal .btn-dropdown {
```
Does class `dropdown--full-width` work instead?

spadgett commented on frontend/public/components/node.tsx:
```diff
@@ -54,9 +55,10 @@ const Header = props => {
+    <ColHead {...props} className="col-md-3 col-sm-4 col-xs-8" sortField="metadata.name">Node Name</ColHead>
+    <ColHead {...props} className="col-md-2 col-sm-2 col-xs-4" sortFunc="nodeReadiness">Status</ColHead>
+    <ColHead {...props} className="col-md-3 col-sm-3 hidden-xs" sortField="metadata.annotations['machine.openshift.io/machine']">Machine</ColHead>
+    <ColHead {...props} className="col-md-1 col-sm-3 hidden-xs">Taints</ColHead>
```
More recent designs have removed taints from the node table. We should go ahead and rollback this change.

nicolethoen commented:
Updated screen shots:

![Screen Shot 2019-03-22 at 2 17 49 PM](https://user-images.githubusercontent.com/26305082/54844512-7e16fe00-4cad-11e9-9fbb-41b61729ffb8.png)
![Screen Shot 2019-03-22 at 2 17 11 PM](https://user-images.githubusercontent.com/26305082/54844517-7f482b00-4cad-11e9-8698-20e468d6a03c.png)

spadgett commented on frontend/public/components/factory/modal.tsx:
```diff
@@ -30,7 +30,7 @@ export const createModalLauncher: CreateModalLauncher = (Component) => (props) =
           contentLabel="Modal"
           onRequestClose={closeModal}
-          className="modal-dialog"
+          className={`modal-dialog ${props.modalClassName}`}
```
```suggestion
          className={classNames('modal-dialog', props.modalClassName)}
```

spadgett replied:
In this case, I would simply assign the class to a `div` inside the dialog. You should be able to achieve the same result without needing this extra prop.

nicolethoen replied:
I will look into this to be sure, but I was trying to utilize the bootstrap class modal-lg which I think needs to be at the same level as modal-dialog to properly adjust the size of the modal.

spadgett replied:
OK makes sense

spadgett commented on frontend/public/components/factory/modal.tsx:
```diff
@@ -87,6 +87,7 @@ export type CreateModalLauncherProps = {
 export type ModalComponentProps = {
   cancel: (e?: Event) => void;
   close: (e?: Event) => void;
+  modalClassName?: string;
```
Most components use simply `className` as the prop name

spadgett commented on frontend/public/components/modals/index.ts:
```diff
@@ -54,3 +59,9 @@ export const clusterChannelModal = (props) => import('./cluster-channel-modal' /
   .then(m => m.clusterUpdateModal(props));
+
+export const taintUpdateModal = (props) => import('./taint-update-modal' /* webpackChunkName: "taint-update-modal" */)
+  .then(m => m.taintUpdateModal(props));
```
nit: I'd probably call these `taintsModal` and `tolerationsModal` (like `labelsModal` above). `clusterUpdateModal` has "update" in the name because you're upgrading the cluster

spadgett commented on frontend/public/components/modals/taint-update-modal.tsx:
```diff
@@ -0,0 +1,140 @@
+import {createModalLauncher, ModalBody, ModalComponentProps, ModalSubmitFooter, ModalTitle} from '../factory';
+import * as _ from 'lodash-es';
+
+class TaintUpdateModal extends PromiseComponent {
```
Again, I'd suggest `TaintsModal`

spadgett commented on frontend/public/components/modals/taint-update-modal.tsx:
```diff
@@ -0,0 +1,140 @@
+  _submit = (e) => {
+    e.preventDefault();
+
+    // We just throw away any rows where the key is blank
```
This comment is no longer accurate.

spadgett commented on frontend/public/components/modals/taint-update-modal.tsx:
```diff
@@ -0,0 +1,140 @@
+    e.preventDefault();
+
+    // We just throw away any rows where the key is blank
+    const taints =this.state.taints;
```
```suggestion
    const { taints } = this.state;
```

or just reference `this.state.taints` below since you aren't using destructuring for props

spadgett commented on frontend/public/components/modals/taint-update-modal.tsx:
```diff
@@ -0,0 +1,140 @@
+
+    this.handlePromise(
+      k8sPatch(this.props.resourceKind, this.props.resource, patch)
+    ).then(this.props.close, () => {});
```
```suggestion
    ).then(this.props.close);
```

spadgett commented on frontend/public/components/modals/taint-update-modal.tsx:
```diff
@@ -0,0 +1,140 @@
+              <Dropdown
+                className="taint-update-modal__dropdown"
+                dropDownClassName="dropdown--full-width"
+                id={`taint_effect_dropdown_${i}`}
```
`id` isn't needed

spadgett commented on frontend/public/components/modals/taint-update-modal.tsx:
```diff
@@ -0,0 +1,140 @@
+                onChange={(e) => this._change(e, i, 'effect')}
+                selectedKey={c.effect} />
+            </div>
+            <div className="hidden-lg hidden-md col-sm-5 col-xs-5"></div>
```
It's unclear to me why we have the empty divs

nicolethoen replied:
They are helping with responsive formatting. Since the form uses bootstrap grid for formatting, the padding put on the fields makes it so that nesting columns doesn't lay everything out well, so I used blank divs as spacers so everything lines up when there is an odd number of fields on the form.

spadgett replied:
OK, should we be using column offsets instead?

https://getbootstrap.com/docs/3.3/css/#grid-offsetting

nicolethoen replied:
ah I'll look at that closer

spadgett commented on frontend/public/components/modals/taint-update-modal.tsx:
```diff
@@ -0,0 +1,140 @@
+  key: string;
+  value: string;
+  effect: string;
+  new: boolean;
```
I don't see where this is used.

spadgett commented on frontend/public/components/modals/toleration-update-modal.tsx:
```diff
@@ -0,0 +1,161 @@
+  operator: string;
+  value: string;
+  effect: string;
+  new: boolean;
```
```suggestion
  isNew?: boolean;
```

spadgett commented on frontend/public/components/modals/toleration-update-modal.tsx:
```diff
@@ -0,0 +1,161 @@
+
+  constructor(public props: TolerationUpdateModalProps) {
+    super(props);
+    this.state.tolerations = this.props.resource.spec.tolerations;
```
This path will need change change depending on whether it's a pod or other workload.

spadgett commented on frontend/public/components/modals/toleration-update-modal.tsx:
```diff
@@ -0,0 +1,161 @@
+  _submit = (e) => {
+    e.preventDefault();
+
+    // We just throw away any rows where the key is blank
```
Comment no longer accurate

spadgett commented on frontend/public/components/modals/toleration-update-modal.tsx:
```diff
@@ -0,0 +1,161 @@
+    e.preventDefault();
+
+    // We just throw away any rows where the key is blank
+    const tolerations =this.state.tolerations;
```
```suggestion
    const { tolerations } = this.state;
```

spadgett commented on frontend/public/components/deployment-config.tsx:
```diff
@@ -99,6 +99,8 @@ export const DeploymentConfigDetailsList = ({dc}) => {
     {triggers && <dt>Triggers</dt>}
     {triggers && <dd>{triggers}</dd>}
+    <dt>Tolerations</dt>
+    <dd><a className="co-m-modal-link" onClick={Kebab.factory.ModifyToleration(DeploymentConfigModel, dc).callback}>{pluralize(_.size(dc.spec.template.spec.tolerations), 'Toleration')}</a></dd>
```
I think it would make sense to add this to `ResourceSummary` as an option (off by default). Similar to node selector, annotations, etc.

https://github.com/spadgett/console/blob/e884e44bdf55005adba898f50f73e5bf95493e4f/frontend/public/components/utils/details-page.tsx#L38-L39

We might need to add a `tolerationsPath` prop that defaults to `spec.template.spec.tolerations`. The path could be passed through to the dialog.

nicolethoen replied:
I don't think I need to, I can get it to work without doing that. Plus I think trying to pass a new prop to the modal would involve changing the KebabAction type and would have larger ramifications.

spadgett commented on frontend/public/components/modals/index.ts:
```diff
@@ -54,3 +59,9 @@ export const clusterChannelModal = (props) => import('./cluster-channel-modal' /
 export const clusterUpdateModal = (props) => import('./cluster-update-modal' /* webpackChunkName: "cluster-update-modal" */)
   .then(m => m.clusterUpdateModal(props));
+
+export const taintModal = (props) => import('./taint-modal' /* webpackChunkName: "taint-modal" */)
```
It should be `taintsModal` and `tolerationsModal` (plural) here and throughout. This is consistent with `annotationsModal` and `labelsModal`

spadgett commented on frontend/public/components/modals/toleration-modal.tsx:
```diff
@@ -0,0 +1,170 @@
+  };
+
+  _isEditable = (t) => {
+    return !_.isEqual(this.props.resourceKind.kind,'Pod') || t.isNew;
```
```suggestion
    return this.props.resourceKind.kind === 'Pod' || t.isNew;
```

spadgett commented on frontend/public/components/modals/toleration-modal.tsx:
```diff
@@ -0,0 +1,170 @@
+            </div>
+            <div className="col-md-2 col-sm-5 col-xs-5 toleration-modal__field">
+              <div className="hidden-md hidden-lg text-secondary text-uppercase">Operator</div>
+              {this._isEditable(t) &&
```
Prefer tenary conditions so you don't need to repeat the logic

```
{ this._isEditable(t) ? /* ... */ : /* ... */ }
```

spadgett commented on frontend/public/components/modals/toleration-modal.tsx:
```diff
@@ -0,0 +1,170 @@
+              <Dropdown
+                className="toleration-modal__dropdown"
+                dropDownClassName="dropdown--full-width"
+                id={`toleration_effect_dropdown_${i}`}
```
Remove unused `id`

nicolethoen commented:
@spadgett 
I think Ive addressed all your comments, but there is a bug I'm working through related to the title of the dropdowns which makes it appear as though the wrong item is selected when you delete a row. That should be all that's left so maybe I'll grab another set of eyes in the morning to get that taken care of ASAP.

spadgett commented:
Dropdowns are getting cut off in the modal:

<img width="372" alt="Screen Shot 2019-03-26 at 8 32 41 AM" src="https://user-images.githubusercontent.com/1167259/54997491-32609f00-4fa2-11e9-8392-fc96237d4dff.png">

Default should be an empty row if the node doesn't have any taints:

<img width="924" alt="Screen Shot 2019-03-26 at 8 35 43 AM" src="https://user-images.githubusercontent.com/1167259/54997519-41475180-4fa2-11e9-9cfa-f39a6007cc3f.png">

At the very least, we need a message saying "No taints." But a new row is more consistent with other editors.

The remove button seems misplaced at mobile:
<img width="356" alt="console-8ffcfd784-75snm · Details · OKD 2019-03-26 08-38-21" src="https://user-images.githubusercontent.com/1167259/54997633-8bc8ce00-4fa2-11e9-8336-8de3e68bb7e6.png">

spadgett commented:
Let's open follow up issues for the dropdown getting cut off and the positioning of the remove button at mobile.

spadgett commented:
/retest

spadgett commented:
/hold

nicolethoen commented:
/lgtm

spadgett commented:
@nicolethoen thanks for testing my updates!

spadgett commented:
/retest
/lgtm
/hold cancel

spadgett commented:
/retest

spadgett commented:
/retest

spadgett commented:
/retest

spadgett commented:
/retest

spadgett commented:
/retest

nicolethoen commented:
/retest

spadgett commented:
/retest

spadgett commented:
/retest

spadgett commented:
/retest

spadgett commented:
/retest

spadgett commented:
/retest

nicolethoen commented:
/retest

spadgett commented:
/retest

spadgett commented:
/retest
